File: src/07.greedy/01.arrayGreedy/10.179.go
package array_greedy

import (
	"sort"
	"strconv"
)

func LargestNumber(nums []int) string {
	sort.Slice(nums, func(i, j int) bool {
		a := strconv.Itoa(nums[i])
		b := strconv.Itoa(nums[j])
		return b+a < a+b
	})

	if nums[0] == 0 {
		return "0"
	}

	res := ""

	for _, v := range nums {
		res += strconv.Itoa(v)
	}

	return res
}

File: src/07.greedy/01.arrayGreedy/10.179.ts
export function largestNumber(nums: number[]): string {
    // const strs = nums.map((n) => n.toString());

    // /**
    //  * x = "3", y = "30"
    //  * x + y = "330"
    //  * y + x = "303"
    //  *
    //  * if:  x + y > y + x  =>  x 应在 y 前面 保证最大
    //  */

    // strs.sort((a, b) => {
    //     return Number(b + a) - Number(a + b);
    // });

    // // 存在前导 0
    // if (strs[0] === "0") {
    //     return "0";
    // }

    // return strs.join("");

    nums.sort((a, b) => {
        // 1: 交换顺序, 保证最大, -1: 顺序不变
        return `${b}${a}` > `${a}${b}` ? 1 : -1;
    });

    if (nums[0] === 0) {
        return "0";
    } else {
        return nums.join("");
    }
}
